streams/shared/src/main/scala/zio/stream/experimental/ZTransducer.scala

```diff
@@ -94,10 +100,66 @@ object ZTransducer {
     }
 
   /**
-   * Creates a sink accumulating incoming values into a list of maximum size `n`.
+   * Creates a transducer accumulating incoming values into lists of maximum size `n`.
    */
   def collectAllN[I](n: Long): ZTransducer[Any, Nothing, I, List[I]] =
-    foldUntil[I, List[I]](Nil, n)((list, element) => element :: list).map(_.reverse).filter(_.nonEmpty)
+    foldUntil[I, List[I]](Nil, n)((list, element) => element :: list).map(_.reverse)
+
+  /**
+   * Creates a transducer accumulating incoming values into maps of up to `n` keys. Elements
+   * are mapped to keys using the function `key`; elements mapped to the same key will
+   * be merged with the function `f`.
+   */
+  def collectAllToMapN[K, I](n: Long)(key: I => K)(f: (I, I) => I): ZTransducer[Any, Nothing, I, Map[K, I]] =
```

Comment:
Hmm, if `costFn` had access to `O` in `foldWeighted`, we could probably write this in terms of it.

Reply:
Yeah. That's what I thought as well. Do you think that'd be a good addition? A `costFn` of the form `(I, O) => Long` could allow for more sophisticated forms of cost calculation.

Reply:
Yes, definitely. The most powerful version can then power all other combinators. ZIO is full of the same kind of thing.